src/array.jl

```diff
@@ -54,7 +58,10 @@ function unsafe_free!(a::CuArray)
     ptr = pointer(a)
     if isvalid(ptr.ctx)
         @trace("Finalizing CuArray object at $(Base.pointer_from_objref(a))")
-        Mem.free(ptr)
+        if Mem.refcount(ptr) == 1
```

Comment:
why not put this in `Mem.free`? Then you only need an additional `Mem.retain` to be called from the `CuArray` constructors.